### PR TITLE
feat: add go routines to init command

### PR DESCRIPTION
**Pull request by converge**
Signed-off-by: João Vanzuita <joao@kubeshop.io>

### CODE REVIEW

File: cmd/init.go [hunk 1]
```diff
@@ -54,126 +72,144 @@ to quickly create a Cobra application.`,
...
+
+			if len(awsFlags.AssumeRole) > 0 {
+				log.Println("calling assume role")
+				err := aws.AssumeRole(awsFlags.AssumeRole)
+				if err != nil {
+					log.Println(err)
+					//return err
+				}
+				log.Printf("assuming new AWS credentials based on role %q", awsFlags.AssumeRole)
 			}
```
Comment (6za) [inline cmd/init.go:88]: I believe for safety, this step should be kept serial.. as I believe the bucket creation step will use this information to create a bucket with the role in play.

**Comment by 6za** [comment]
Reviewing the PR for possible race conditions.. as this may impact installer behavior.

File: internal/downloadManager/download.go [hunk 2]
```diff
@@ -30,116 +31,135 @@ func DownloadTools(config *configs.Config) error {
...
-	extractFileFromTarGz(
-		helmTarDownload,
-		fmt.Sprintf("%s-%s/helm", config.LocalOs, config.LocalArchitecture),
-		config.HelmClientPath,
-	)
-	err = os.Chmod(config.HelmClientPath, 0755)
-	if err != nil {
-		return err
-	}
+	go func() {
```
Comment (6za) [inline internal/downloadManager/download.go:112]: Maybe we should use named functions here, to make tis code better to read, also cleaner. 

This function is the one most improved by been parallelized.

File: cmd/init.go [hunk 3]
```diff
@@ -54,126 +72,144 @@ to quickly create a Cobra application.`,
...
+
+		go func() {
+			//! tracker 7
+			log.Printf("cloning and detokenizing the metaphor-template repository")
+			prepareKubefirstTemplateRepo(config, viper.GetString("gitops.owner"), "metaphor", "", viper.GetString("template.tag"))
+			log.Println("clone and detokenization of metaphor-template repository complete")
+			progressPrinter.IncrementTracker("step-metaphor", 1)
+
+			wg.Done()
+		}()
```
Comment (6za) [inline cmd/init.go:210]: This step was removed in current iterations of init in construction.

File: cmd/init.go [hunk 4]
```diff
@@ -54,126 +72,144 @@ to quickly create a Cobra application.`,
...
+
+		wg.Wait()
```
Comment (6za) [inline cmd/init.go:212]: Did you checked if `viper.WriteConfig()` and viper is thread safe? As we have several thread updating values on it. Do we need any mutex protection?

Comment (6za) [reply cmd/init.go:212]: Maybe this could help for details: https://github.com/spf13/viper/issues/19

**Comment by 6za** [comment]
I really like the direction of this PR, I just want we ensure we are aware, that we need to ensure:
- Viper is thread-safe to allow the parallel steps to no break as create is heavily dependent of `init`final kubefirst file. 
- Code is kept easy to read, thread(and similar like go routines, add some complexities to manage "order")

**Comment by converge** [comment]
as mentioned https://github.com/spf13/viper/issues/19 , viper is thread unsafe. I'll rewrite the PR at some point to speed up init.

> Code is kept easy to read, thread(and similar like go routines, add some complexities to manage "order")

as you guys are changing a lot of the structure, I don't want to add new functions at this point. I would might need to refactor a bit, and this will cause merging conflicts on you guys current PRs.

**Pull request closed without merging**